File: AmpedBiz/AmpedBiz.Service/Dto/PurchaseOrder.cs
using AmpedBiz.Common.CustomTypes;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;

namespace AmpedBiz.Service.Dto
{
    public enum PurchaseOrderStatus
    {
        Created = 1,
        Submitted = 2,
        Approved = 3,
        Completed = 4,
        Cancelled = 5
    }

    public enum PurchaseOrderAggregate
    {
        Items = 1,
        Payments = 2,
        Receipts = 3
    }

    public class PurchaseOrder
    {
        public Guid Id { get; set; }

        public string ReferenceNumber { get; set; }

        public string VoucherNumber { get; set; }

        public Guid UserId { get; set; }

        public Lookup<string> PaymentType { get; set; }

        public Lookup<string> Shipper { get; set; }

        public Lookup<Guid> Supplier { get; set; }

        public decimal TaxAmount { get; set; }

        public decimal ShippingFeeAmount { get; set; }

        public decimal SubTotalAmount { get; set; }

        public decimal TotalAmount { get; set; }

        public decimal PaidAmount { get; set; }

        public PurchaseOrderStatus Status { get; set; }

        public DateTime? ExpectedOn { get; set; }

        public Lookup<Guid> CreatedBy { get; set; }

        public DateTime? CreatedOn { get; set; }

        public DateTime? RecreatedOn { get; set; }

        public Lookup<Guid> RecreatedBy { get; set; }

        public Lookup<Guid> SubmittedBy { get; set; }

        public DateTime? SubmittedOn { get; set; }

        public Lookup<Guid> ApprovedBy { get; set; }

        public DateTime? ApprovedOn { get; set; }

        public Lookup<Guid> PaidBy { get; set; }

        public DateTime? PaidOn { get; set; }

        public Lookup<Guid> CompletedBy { get; set; }

        public DateTime? CompletedOn { get; set; }

        public Lookup<Guid> CancelledBy { get; set; }

        public DateTime? CancelledOn { get; set; }

        public string CancellationReason { get; set; }

        public IEnumerable<PurchaseOrderItem> Items { get; set; } = new Collection<PurchaseOrderItem>();

        public IEnumerable<PurchaseOrderPayment> Payments { get; set; } = new Collection<PurchaseOrderPayment>();

        public IEnumerable<PurchaseOrderReceipt> Receipts { get; set; } = new Collection<PurchaseOrderReceipt>();

        public IEnumerable<PurchaseOrderReceivable> Receivables { get; set; } = new Collection<PurchaseOrderReceivable>();

        public Dictionary<PurchaseOrderStatus, string> AllowedTransitions { get; set; }

        public StageDefenition<PurchaseOrderStatus, PurchaseOrderAggregate> Stage { get; set; }
    }

    public class PurchaseOrderPageItem
    {
        public Guid Id { get; set; }

        public string ReferenceNumber { get; set; }

        public string VoucherNumber { get; set; }

        public string Supplier { get; set; }

        public string Status { get; set; }

        public string CreatedBy { get; set; }

        public DateTime? CreatedOn { get; set; }

        public string SubmittedBy { get; set; }

        public DateTime? SubmittedOn { get; set; }

        public string PaidBy { get; set; }

        public DateTime? PaidOn { get; set; }

        public decimal? TotalAmount { get; set; }
    }
}

File: AmpedBiz/AmpedBiz.Web/src/common/models/purchase-order.ts
import { Lookup } from '../custom_types/lookup';
import { Measure } from "./measure";
import { StageDefinition } from './stage-definition';
import { UnitOfMeasure } from "./unit-of-measure";

export const purchaseOrderEvents = {
  item: {
    add: 'purchase-order-item-add',
    added: 'purchase-order-item-added',
    deleted: 'purchase-order-item-deleted',
  },
  payment: {
    add: 'purchase-order-payment-add',
    added: 'purchase-order-payment-added',
    deleted: 'purchase-order-payment-deleted',
  },
  receipts: {
    add: 'purchase-order-receipt-add',
    added: 'purchase-order-receipt-added',
    deleted: 'purchase-order-receipt-deleted',
    receive: 'purchase-order-receipt-receive',
    received: 'purchase-order-receipt-received',
  },
}

export enum PurchaseOrderStatus {
  created = 1,
  submitted = 2,
  approved = 3,
  completed = 4,
  cancelled = 5
}

export enum PurchaseOrderAggregate {
  items = 1,
  payments = 2,
  receipts = 3
}

export type PurchaseOrder = {
  id?: string;
  referenceNumber?: string;
  voucherNumber?: string;
  branch?: Lookup<string>;
  user?: Lookup<string>;
  supplier?: Lookup<string>;
  shipper?: Lookup<string>;
  paymentType?: string;
  taxAmount?: number;
  shippingFeeAmount?: number;
  discountAmount?: number;
  subTotalAmount?: number;
  totalAmount?: number;
  paymentAmount?: number;
  status?: PurchaseOrderStatus;
  expectedOn?: Date;
  createdBy?: Lookup<string>;
  createdOn?: Date;
  recreatedOn?: Date;
  recreatedBy?: Lookup<string>;
  submittedBy?: Lookup<string>;
  submittedOn?: Date;
  closedBy?: Lookup<string>;
  closedOn?: Date;
  cancelledBy?: Lookup<string>,
  cancelledOn?: Date,
  cancellationReason?: string,
  items?: PurchaseOrderItem[];
  payments?: PurchaseOrderPayment[];
  receipts?: PurchaseOrderReceipt[];
  receivables?: PurchaseOrderReceivable[];
  stage?: StageDefinition<PurchaseOrderStatus, PurchaseOrderAggregate>;
}

export type PurchaseOrderPageItem = {
  id?: string;
  supplier?: string;
  status?: PurchaseOrderStatus;
  createdBy?: string;
  createdOn?: Date;
  submittedBy?: string;
  submittedOn?: Date;
  paymentBy?: string;
  paymentOn?: Date;
  total?: string;
}

export type PurchaseOrderItem = {
  id?: string;
  product?: Lookup<string>;
  unitOfMeasures?: UnitOfMeasure[];
  quantity?: Measure;
  standard?: Measure;
  unitCostAmount?: number;
  totalCostAmount?: number;
}

export type PurchaseOrderPayment = {
  id?: string;
  purchaseOrderId?: string;
  paymentBy?: Lookup<string>;
  paymentOn?: Date;
  paymentAmount?: number;
  paymentType?: Lookup<string>;
}

export type PurchaseOrderReceipt = {
  id?: string;
  purchaseOrderId?: string;
  batchNumber?: string;
  receivedBy?: Lookup<string>;
  receivedOn?: Date;
  expiresOn?: Date;
  product?: Lookup<string>;
  quantity?: Measure;
  standard?: Measure;
}

export type PurchaseOrderReceivable = {
  purchaseOrderId?: string;
  product?: Lookup<string>;
  unitOfMeasures?: UnitOfMeasure[];
  orderedQuantity?: number;
  receivedQuantity?: number;
  receivableQuantity?: number;
  standard?: Measure;
  receiving: PurchaseOrderReceiving;
}

export type PurchaseOrderReceiving = {
  batchNumber?: string;
  receivedBy?: Lookup<string>;
  receivedOn?: Date;
  expiresOn?: Date;
  quantity?: Measure;
  standard?: Measure;
}

export type PurchaseOrderPayable = {
  id?: string;
  purchaseOrderId?: string;
  paymentOn?: Date;
  paymentBy?: Lookup<string>;
  paymentType?: Lookup<string>;
  taxAmount?: number;
  shippingFeeAmount?: number;
  discountAmount?: number;
  subTotalAmount?: number;
  totalAmount?: number;
  balanceAmount?: number;
  paidAmount?: number;
  paymentAmount?: number;
}

export type Voucher = {
  id?: string;
  supplierName?: string;
  referenceNumber?: string;
  voucherNumber?: string;
  approvedByName?: string;
  approvedOn?: Date;
  paymentTypeName?: string;
  taxAmount?: number;
  shippingFeeAmount?: number;
  subTotalAmount?: number;
  totalAmount?: number;
  items: VoucherItem[];
}

export type VoucherItem = {
  id?: string;
  product?: Lookup<string>;
  quantity?: Measure;
  unitCostAmount?: number;
  totalCostAmount?: number;
}

export type PurchaseOrderReportPageItem = {
  id?: string;
  createdOn?: Date;
  branchName?: string;
  supplierName?: string;
  voucherNumber?: string;
  createdByName?: string;
  approvedByName?: string;
  status?: PurchaseOrderStatus;
  totalAmount?: number;
  paidAmount?: number;
  balanceAmount?: number;
}
